miio/gateway/gateway.py

```diff
@@ -0,0 +1,335 @@
+    * get_corridor_on_time
+    * set_corridor_light ["off"]
+    * get_corridor_light -> "on"
+
+    * set_default_sound
+    * set_doorbell_push, get_doorbell_push ["off"]
+    * set_doorbell_volume [100], get_doorbell_volume
+    * set_gateway_volume, get_gateway_volume
+    * set_clock_volume
+    * set_clock
+    * get_sys_data
+    * update_neighbor_token [{"did":x, "token":x, "ip":x}]
+
+    ## property getters
+    * ctrl_device_prop
+    * get_device_prop_exp [[sid, list, of, properties]]
+
+    ## scene
+    * get_lumi_bind ["scene", <page number>] for rooms/devices"""
+
+    def __init__(
+        self,
+        ip: str = None,
+        token: str = None,
+        start_id: int = 0,
+        debug: int = 0,
+        lazy_discover: bool = True,
+    ) -> None:
+        super().__init__(ip, token, start_id, debug, lazy_discover)
+
+        from . import (
+            Alarm,
+            Radio,
+            Zigbee,
+            Light,
+        )
+
+        self._alarm = Alarm(parent=self)
+        self._radio = Radio(parent=self)
+        self._zigbee = Zigbee(parent=self)
+        self._light = Light(parent=self)
+        self._devices = {}
+        self._info = None
+        self._subdevice_model_map = None
+
+    def _get_subdevice_model_map(self):
+        if self._subdevice_model_map is None:
+            filedata = open(os.path.dirname(__file__) + "/devices/subdevices.yaml", "r")
+            self._subdevice_model_map = yaml.safe_load(filedata)
+        return self._subdevice_model_map
+
+    def _get_unknown_model(self):
+        self._get_subdevice_model_map()
+
+        for model_info in self._subdevice_model_map:
+            if model_info.get("type_id") == -1:
+                return model_info
+
+    @property
+    def alarm(self) -> "GatewayAlarm":  # noqa: F821
+        """Return alarm control interface."""
+        # example: gateway.alarm.on()
+        return self._alarm
+
+    @property
+    def radio(self) -> "GatewayRadio":  # noqa: F821
+        """Return radio control interface."""
+        return self._radio
+
+    @property
+    def zigbee(self) -> "GatewayZigbee":  # noqa: F821
+        """Return zigbee control interface."""
+        return self._zigbee
+
+    @property
+    def light(self) -> "GatewayLight":  # noqa: F821
+        """Return light control interface."""
+        return self._light
+
+    @property
+    def devices(self):
+        """Return a dict of the already discovered devices."""
+        return self._devices
+
+    @property
+    def model(self):
+        """Return the zigbee model of the gateway."""
+        # Check if catch already has the gateway info, otherwise get it from the device
+        if self._info is None:
+            self._info = self.info()
+        return self._info.model
+
+    @command()
+    def discover_devices(self):
+        """
+        Discovers SubDevices
+        and returns a list of the discovered devices.
+        """
+
+        from .devices import SubDeviceInfo
```

Comment:
This should be moved to the beginning of the file.

Reply:
That is not possible:
Because a gateway can be supplied to the init of the `SubDevice` class, there is a argument type `Gateway`.
Therefore at the top of the `devices.py` file the `Gateway` class is imported from the "gateway.py" file.

Therefore somthing from "devices.py" can only be imported from the "gateway.py" file after the line in which the  `Gateway` class is defined, otherwise you create a cyclic import (a loop that will endlessly jump between the files to try and import something from one another. (same thing holds for the `GatewayException` class).

It can therefore only be moved up (but below the `GatewayException` class) if the type definition is removed from the `__init__` of the `SubDevice` class.

Do you want to keep it like this, or remove the type definition?

Reply:
Ah, gotcha! You can remove the import (in `subdevice.py`), and simply use the class name in quotation marks and it will work just as well as imported type hint. Would that work?

Reply:
yea that will probably work, will change it and test it now.

Reply:
I tested it and it works, just pushed the commit.